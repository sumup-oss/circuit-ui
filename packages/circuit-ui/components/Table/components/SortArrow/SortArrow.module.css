.base {
  position: absolute;
  top: 50%;
  left: 0;
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: var(--cui-spacings-giga);
  height: 36px;
  padding: 2px 4px;
  margin: 0;
  color: var(--cui-fg-accent);
  cursor: pointer;
  outline: 0;
  background: none;
  border: 0;
  opacity: 0;
  transform: translateY(-50%);
  transition: opacity var(--cui-transitions-default);
}

.base:focus {
  opacity: 1;
}

.base:focus::-moz-focus-inner {
  border: 0;
}

.icon {
  margin: -2px 0;
}
