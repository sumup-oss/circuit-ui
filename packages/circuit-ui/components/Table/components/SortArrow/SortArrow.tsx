/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { HTMLAttributes } from 'react';
import { css } from '@emotion/react';
import { ChevronUp, ChevronDown } from '@sumup/icons';

import styled, { StyleProps } from '../../../../styles/styled';
import { hideVisually } from '../../../../styles/style-mixins';
import { Direction } from '../../types';

interface SortArrowProps extends HTMLAttributes<HTMLButtonElement> {
  direction?: Direction;
  label: string;
}

const baseStyles = ({ theme }: StyleProps) => css`
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: 36px;
  width: ${theme.spacings.giga};
  position: absolute;
  left: 0;
  top: 50%;
  opacity: 0;
  transform: translateY(-50%);
  transition: opacity ${theme.transitions.default};
  color: var(--cui-fg-accent);
  border: 0;
  background: none;
  outline: 0;
  padding: 2px 4px;
  margin: 0;
  cursor: pointer;

  &:focus {
    opacity: 1;

    &::-moz-focus-inner {
      border: 0;
    }
  }
`;

const Button = styled.button(baseStyles);

const iconStyles = css`
  margin: -2px 0;
`;

const Label = styled('span')(hideVisually);

/**
 * SortArrow for the Table component. The Table handles rendering it.
 */
const SortArrow = ({
  label,
  direction,
  ...props
}: SortArrowProps): JSX.Element => (
  <Button role="button" title={label} {...props}>
    {direction !== 'ascending' && <ChevronUp size="16" css={iconStyles} />}
    {direction !== 'descending' && <ChevronDown size="16" css={iconStyles} />}
    <Label>{label}</Label>
  </Button>
);

export default SortArrow;
