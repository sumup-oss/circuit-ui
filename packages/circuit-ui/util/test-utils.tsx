/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FunctionComponent, ReactElement } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import '@testing-library/jest-dom/extend-expect';
import { configureAxe } from 'jest-axe';
import { render as renderTest, RenderResult } from '@testing-library/react';
import { renderHook, act as actHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'emotion-theming';
import { light } from '@sumup/design-tokens';

import {
  ComponentsContext,
  defaultComponents,
} from '../components/ComponentsContext/ComponentsContext';

export * from '@testing-library/react';

export type RenderFn<T = any> = (component: ReactElement, ...rest: any) => T;

const WithProviders: FunctionComponent = ({ children }) => (
  <ComponentsContext.Provider value={defaultComponents}>
    <ThemeProvider theme={light}>{children}</ThemeProvider>
  </ComponentsContext.Provider>
);

const render: RenderFn<RenderResult> = (component, options) =>
  renderTest(component, { wrapper: WithProviders, ...options });
const renderToHtml: RenderFn<string> = (component) =>
  renderToStaticMarkup(<WithProviders>{component}</WithProviders>);
const create = (...args: Parameters<RenderFn<RenderResult>>) => {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
};

const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export { create, render, renderToHtml, renderHook, actHook, userEvent, axe };
