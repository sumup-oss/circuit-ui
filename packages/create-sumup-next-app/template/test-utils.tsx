import React, { FunctionComponent } from 'react';
import { configureAxe } from 'jest-axe';
import { ThemeProvider } from '@emotion/react';
import { render as renderTest } from '@testing-library/react';
import { light } from '@sumup/design-tokens';

const WithProviders: FunctionComponent = ({ children }) => (
  <ThemeProvider theme={light}>{children}</ThemeProvider>
);

export const render = (
  component: Parameters<typeof renderTest>[0],
  options: Parameters<typeof renderTest>[1] = {},
) => renderTest(component, { wrapper: WithProviders, ...options });

export const axe = configureAxe({
  rules: {
    // Disable landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export * from '@testing-library/react';
