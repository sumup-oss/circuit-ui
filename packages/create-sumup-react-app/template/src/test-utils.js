import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import { light } from '@sumup/design-tokens';
import PropTypes from 'prop-types';

const AllTheProviders = ({ children }) => (
  <ThemeProvider theme={light}>{children}</ThemeProvider>
);
AllTheProviders.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };
