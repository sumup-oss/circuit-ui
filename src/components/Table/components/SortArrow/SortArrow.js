/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { hideVisually } from 'polished';
import { ChevronUp, ChevronDown } from '@sumup/icons';

import { ASCENDING, DESCENDING } from '../../constants';

const baseStyles = ({ theme }) => css`
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: 40px;
  width: 20px;
  position: absolute;
  left: 0;
  top: 50%;
  opacity: 0;
  transform: translateY(-50%);
  transition: opacity ${theme.transitions.default};
  color: ${theme.colors.p500};
  border: 0;
  background: none;
  outline: 0;
  padding: 0;
  margin: 0;
  cursor: pointer;

  &:focus {
    opacity: 1;

    &::-moz-focus-inner {
      border: 0;
    }
  }
`;

const Wrapper = styled.button(baseStyles);

const iconStyles = () => css`
  margin-top: -7px;
  margin-bottom: -7px;
`;

const Label = styled('span')(hideVisually);

/**
 * @private Arrow component for TableHeader sorting
 */
const SortArrow = ({ label = 'Sort', direction = null, ...props }) => (
  <Wrapper role="button" title={label} {...props}>
    {direction !== ASCENDING && <ChevronUp css={iconStyles} />}
    {direction !== DESCENDING && <ChevronDown css={iconStyles} />}
    <Label>{label}</Label>
  </Wrapper>
);

SortArrow.propTypes = {
  direction: PropTypes.oneOf([ASCENDING, DESCENDING]),
  label: PropTypes.string
};

export default SortArrow;
