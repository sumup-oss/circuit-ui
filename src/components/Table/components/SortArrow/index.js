/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { ReactComponent as ArrowIcon } from './arrow.svg';

import SvgButton from '../../../SvgButton';
import { ASCENDING, DESCENDING } from '../../constants';

const baseStyles = ({ theme }) => css`
  display: flex;
  flex-direction: column;
  height: 10px;
  left: ${theme.spacings.kilo};
  opacity: 0;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  transition: opacity ${theme.transitions.default};
  width: 5px;
`;

const StyledWrapper = styled(SvgButton)`
  ${baseStyles};
`;

const DownArrow = styled(ArrowIcon)`
  margin-top: 2px;
  transform: rotate(180deg);
`;

/**
 * [PRIVATE] Arrow component for TableHeader sorting
 */
const SortArrow = ({ direction }) => (
  <StyledWrapper>
    <Fragment>
      {direction !== DESCENDING && <ArrowIcon />}
      {direction !== ASCENDING && <DownArrow />}
    </Fragment>
  </StyledWrapper>
);

SortArrow.propTypes = {
  direction: PropTypes.oneOf([ASCENDING, DESCENDING])
};

SortArrow.defaultProps = {
  direction: null
};

export default SortArrow;
