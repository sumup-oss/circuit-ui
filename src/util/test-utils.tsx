/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { FunctionComponent } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import '@testing-library/jest-dom/extend-expect';
import { configureAxe } from 'jest-axe';
import { render as renderTest, wait, act } from '@testing-library/react';
import { renderHook, act as actHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'emotion-theming';
import { light } from '@sumup/design-tokens';

import ComponentsContext, {
  defaultComponents,
} from '../components/ComponentsContext';

export type RenderFn = (component: React.ReactElement, ...rest: any) => any;

const WithProviders: FunctionComponent = ({ children }) => (
  <ComponentsContext.Provider value={defaultComponents}>
    <ThemeProvider theme={light}>{children}</ThemeProvider>
  </ComponentsContext.Provider>
);

const renderWithProviders = (renderer: RenderFn): RenderFn => (
  component,
  ...rest
): RenderFn => renderer(<WithProviders>{component}</WithProviders>, rest);

const render: RenderFn = (component, options) =>
  renderTest(component, { wrapper: WithProviders, ...options });
const renderToHtml: RenderFn = renderWithProviders(renderToStaticMarkup);
const create: RenderFn = (...args) => {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
};

const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export {
  create,
  render,
  renderToHtml,
  renderHook,
  act,
  actHook,
  userEvent,
  wait,
  axe,
};
