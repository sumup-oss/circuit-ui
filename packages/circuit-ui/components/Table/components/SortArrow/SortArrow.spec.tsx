/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { describe, expect, it, vi } from 'vitest';

import { render, axe, userEvent, screen } from '../../../../util/test-utils.js';

import SortArrow from './index.js';

describe('SortArrow', () => {
  it('should render with both arrows styles', () => {
    const { container } = render(<SortArrow label="Sort" />);
    expect(container.querySelectorAll('svg')).toHaveLength(2);
  });

  it('should render with ascending arrow styles', () => {
    const { container } = render(
      <SortArrow label="Sort" direction="ascending" />,
    );
    expect(container.querySelectorAll('svg')).toHaveLength(1);
  });

  it('should render with descending arrow styles', () => {
    const { container } = render(
      <SortArrow label="Sort" direction="descending" />,
    );
    expect(container.querySelectorAll('svg')).toHaveLength(1);
  });

  it('should call the onClick callback', async () => {
    const onClick = vi.fn();
    render(<SortArrow label="Sort" onClick={onClick} data-testid="sort" />);
    await userEvent.click(screen.getByTestId('sort'));
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('should have no accessibility violations', async () => {
    const { container } = render(<SortArrow label="Sort" />);
    const actual = await axe(container);
    expect(actual).toHaveNoViolations();
  });
});
