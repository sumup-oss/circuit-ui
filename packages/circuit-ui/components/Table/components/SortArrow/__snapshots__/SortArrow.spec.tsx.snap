// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortArrow Style tests should render with ascending arrow styles 1`] = `
.circuit-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  height: 36px;
  width: 24px;
  position: absolute;
  left: 0;
  top: 50%;
  opacity: 0;
  -webkit-transform: translateY(-50%);
  -moz-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: opacity 120ms ease-in-out;
  transition: opacity 120ms ease-in-out;
  color: var(--cui-fg-accent);
  border: 0;
  background: none;
  outline: 0;
  padding: 2px 4px;
  margin: 0;
  cursor: pointer;
}

.circuit-0:focus {
  opacity: 1;
}

.circuit-0:focus::-moz-focus-inner {
  border: 0;
}

.circuit-1 {
  margin: -2px 0;
}

.circuit-2 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<button
  class="circuit-0"
  role="button"
  title="Sort"
>
  <svg
    class="circuit-1"
    fill="none"
    height="16"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M8 11c-.266 0-.52-.104-.71-.29l-4-4A1.013 1.013 0 0 1 3 6a1 1 0 0 1 1-1c.266 0 .52.104.71.29L8 8.58l3.29-3.29c.19-.186.445-.29.71-.29a1 1 0 0 1 1 1c0 .266-.104.52-.29.71l-4 4c-.19.186-.444.29-.71.29z"
      fill="currentColor"
    />
  </svg>
  <span
    class="circuit-2"
  >
    Sort
  </span>
</button>
`;

exports[`SortArrow Style tests should render with both arrows styles 1`] = `
.circuit-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  height: 36px;
  width: 24px;
  position: absolute;
  left: 0;
  top: 50%;
  opacity: 0;
  -webkit-transform: translateY(-50%);
  -moz-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: opacity 120ms ease-in-out;
  transition: opacity 120ms ease-in-out;
  color: var(--cui-fg-accent);
  border: 0;
  background: none;
  outline: 0;
  padding: 2px 4px;
  margin: 0;
  cursor: pointer;
}

.circuit-0:focus {
  opacity: 1;
}

.circuit-0:focus::-moz-focus-inner {
  border: 0;
}

.circuit-1 {
  margin: -2px 0;
}

.circuit-3 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<button
  class="circuit-0"
  role="button"
  title="Sort"
>
  <svg
    class="circuit-1"
    fill="none"
    height="16"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M8 5c.266 0 .52.104.71.29l4 4c.186.19.29.444.29.71a1 1 0 0 1-1 1c-.265 0-.52-.104-.71-.29L8 7.42l-3.29 3.29c-.19.186-.444.29-.71.29a1 1 0 0 1-1-1c0-.266.104-.52.29-.71l4-4C7.48 5.104 7.734 5 8 5z"
      fill="currentColor"
    />
  </svg>
  <svg
    class="circuit-1"
    fill="none"
    height="16"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M8 11c-.266 0-.52-.104-.71-.29l-4-4A1.013 1.013 0 0 1 3 6a1 1 0 0 1 1-1c.266 0 .52.104.71.29L8 8.58l3.29-3.29c.19-.186.445-.29.71-.29a1 1 0 0 1 1 1c0 .266-.104.52-.29.71l-4 4c-.19.186-.444.29-.71.29z"
      fill="currentColor"
    />
  </svg>
  <span
    class="circuit-3"
  >
    Sort
  </span>
</button>
`;

exports[`SortArrow Style tests should render with descending arrow styles 1`] = `
.circuit-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  height: 36px;
  width: 24px;
  position: absolute;
  left: 0;
  top: 50%;
  opacity: 0;
  -webkit-transform: translateY(-50%);
  -moz-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: opacity 120ms ease-in-out;
  transition: opacity 120ms ease-in-out;
  color: var(--cui-fg-accent);
  border: 0;
  background: none;
  outline: 0;
  padding: 2px 4px;
  margin: 0;
  cursor: pointer;
}

.circuit-0:focus {
  opacity: 1;
}

.circuit-0:focus::-moz-focus-inner {
  border: 0;
}

.circuit-1 {
  margin: -2px 0;
}

.circuit-2 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<button
  class="circuit-0"
  role="button"
  title="Sort"
>
  <svg
    class="circuit-1"
    fill="none"
    height="16"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M8 5c.266 0 .52.104.71.29l4 4c.186.19.29.444.29.71a1 1 0 0 1-1 1c-.265 0-.52-.104-.71-.29L8 7.42l-3.29 3.29c-.19.186-.444.29-.71.29a1 1 0 0 1-1-1c0-.266.104-.52.29-.71l4-4C7.48 5.104 7.734 5 8 5z"
      fill="currentColor"
    />
  </svg>
  <span
    class="circuit-2"
  >
    Sort
  </span>
</button>
`;
